usaspending_api/recipient/delta_models/recipient_lookup.py

```diff
@@ -1,30 +1,686 @@
+            ultimate_parent_unique_ide AS parent_duns,
+            NULL AS parent_legal_business_name,
+            ultimate_parent_uei AS parent_uei,
+            NULL AS state,
+            NULL AS zip4,
+            NULL AS zip5,
+            update_date,
+            NULL AS alternate_names,
+            ROW_NUMBER() OVER (PARTITION BY ultimate_parent_uei, ultimate_parent_unique_ide ORDER BY update_date DESC NULLS LAST) AS row_num
+        FROM raw.sam_recipient
+        WHERE COALESCE(ultimate_parent_uei, ultimate_parent_unique_ide) IS NOT NULL AND ultimate_parent_legal_enti IS NULL
+    ),
+    latest_tx_parent_no_name AS (
+        SELECT
+            8 AS priority,
+            parent_recipient_hash AS recipient_hash,
+            duns_parent_recipient_hash AS duns_recipient_hash,
+            NULL AS legal_business_name,
+            ultimate_parent_unique_ide AS duns,
+            ultimate_parent_uei AS uei,
+            CONCAT(source, '-parent') AS source,
+            NULL AS address_line_1,
+            NULL AS address_line_2,
+            NULL AS business_types_codes,
+            NULL AS city,
+            NULL AS congressional_district,
+            NULL AS country_code,
+            ultimate_parent_unique_ide AS parent_duns,
+            NULL AS parent_legal_business_name,
+            ultimate_parent_uei AS parent_uei,
+            NULL AS state,
+            NULL AS zip4,
+            NULL AS zip5,
+            action_date AS update_date,
+            NULL AS alternate_names,
+            ROW_NUMBER() OVER (PARTITION BY parent_recipient_hash ORDER BY action_date DESC, is_fpds, transaction_unique_id) AS row_num
+        FROM temporary_transaction_recipients_view
+        WHERE COALESCE(ultimate_parent_uei, ultimate_parent_unique_ide) IS NOT NULL AND ultimate_parent_legal_enti IS NULL
+    ),
+    latest_tx_no_fabs_fpds AS (
+        SELECT
+            9 AS priority,
+            recipient_hash,
+            duns_recipient_hash,
+            awardee_or_recipient_legal AS legal_business_name,
+            NULL AS duns,
+            NULL AS uei,
+            source,
+            address_line_1,
+            address_line_2,
+            NULL AS business_types_codes,
+            city,
+            congressional_district,
+            country_code,
+            ultimate_parent_unique_ide AS parent_duns,
+            ultimate_parent_legal_enti AS parent_legal_business_name,
+            ultimate_parent_uei AS parent_uei,
+            state,
+            zip4,
+            zip5,
+            action_date AS update_date,
+            NULL AS alternate_names,
+            ROW_NUMBER() OVER (PARTITION BY recipient_hash ORDER BY action_date DESC, is_fpds, transaction_unique_id) AS row_num
+        FROM temporary_transaction_recipients_view
+        WHERE COALESCE(uei, awardee_or_recipient_uniqu) IS NULL
+    ),
+    union_all AS (
+        SELECT * FROM latest_duns_sam WHERE row_num = 1
+        UNION ALL
```

Comment:
Is UNION ALL what we want here? It collects all the rows from each step and doesn't preserve priority (i.e. if the SAME row is surfaced by priority 1 step and priority 2 step, it's not just priority 1 that remains, but both of them).

I guess a straight UNION, even if supported by Spark (not sure if they do both UNION and UNION ALL differently) wouldn't work if we're adding the priority in though. So maybe there's no way around it.

Reply:
The thought I had, not completely thought-out, was to `UNION` the datasets in reverse priority order, so last-one-wins, replacing the same rows from earlier datasets. Or something like that. But I guess ALL rows would have to be equal to overwrite, and it's only some of them that are. 
So maybe first collecting them in 1 `UNION ALL` dataset, and then using `PARTITION` to get the _1st_ row (based on e.g. the `ORDER BY priority` column) ... would achieve the same effect. --> which you have below in the `union_all_priority`. 

And PARTITIONing on `recipient_hash` (which I _think_ will _always_ have a value in the above CTEs) is the same as sequential `MERGE` statements with `recipient_hash` as the merge-matching-key. So I think all is good here 👍

Reply:
For now this has been updated to limit with `WHERE row_num_union = 1`, however, I plan to create a bug ticket since this functionality is actually a bug since it limits the recipients that we use to generate alternate_names and also remove old recipients that are still aligned to their DUNS and not UEI.